File: Grammars/Dice/DiceRoll.vb
Imports System.Xml.Serialization
Imports System.Collections.ObjectModel
Imports System.ComponentModel
Imports System.Runtime.CompilerServices

Namespace Dice
    <System.Xml.Serialization.XmlRoot("DiceRoller", IsNullable:=False)>
    Public Class DiceRoll
        Inherits BaseGrammar

        Private Const ASSIGN_OPERATOR As String = ":="
        Private Const ROLL_FUNCTION_PARAMETER As String = "RollFunction"
        Private Const COMMENT_CHARACTER As String = "#"

        Private WithEvents _calculator As New Calculator(Me)
        Private _results As New Dictionary(Of Int32, String)

        Public Class RollFunction
            Implements INotifyPropertyChanged
            Implements IDataErrorInfo

            Public Sub New()
            End Sub

            Public Event PropertyChanged(sender As Object, e As PropertyChangedEventArgs) Implements INotifyPropertyChanged.PropertyChanged

            Private _name As String
            Private _value As String

            <XmlAttribute("name")>
            Public Property Name As String
                Get
                    Return _name
                End Get
                Set(value As String)
                    If _name <> value Then
                        _name = value
                        OnPropertyChanged()
                    End If
                End Set
            End Property
            <XmlText>
            Public Property Value As String
                Get
                    Return _value
                End Get
                Set(value As String)
                    If _value <> value Then
                        _value = value
                        OnPropertyChanged()
                    End If
                End Set
            End Property

            Public ReadOnly Property [Error] As String Implements IDataErrorInfo.Error
                Get
                    Return String.Empty
                End Get
            End Property

            Default Public ReadOnly Property Item(columnName As String) As String Implements IDataErrorInfo.Item
                Get
                    Select Case columnName
                        Case "Name" : If String.IsNullOrWhiteSpace(Name) Then Return "Name is required."
                        Case "Value" : If String.IsNullOrWhiteSpace(Value) Then Return "Function is required."
                    End Select
                    Return String.Empty
                End Get
            End Property

            Protected Overridable Sub OnPropertyChanged(<CallerMemberNameAttribute> Optional ByVal propertyName As String = Nothing)
                RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
            End Sub
        End Class

        <XmlElement("function")>
        Property RollFunctions As New ObservableCollection(Of RollFunction)

        <XmlElement("supportsMaxLength")>
        Public Overrides Property SupportsMaxLength As Boolean
            Get
                Return False
            End Get
            Set(value As Boolean)
            End Set
        End Property

        Public Overrides Function Analyze() As String
            Return String.Empty
        End Function

        Protected Overrides Function GenerateName() As String
            Dim lines As New List(Of String)(Split(GetRollFunction, Chr(&HA)))
            Dim lineNumber As Int32 = 1

            _results.Clear()

            For Each line As String In lines
                line = line.Trim()
                If Not String.IsNullOrWhiteSpace(line) AndAlso Not line.StartsWith(COMMENT_CHARACTER) Then
                    Dim parts() As String = Split(line, ASSIGN_OPERATOR)
                    If parts.Count = 1 Then
                        _results.Add(lineNumber, _calculator.Evaluate(line).ToString)
                    Else
                        If GetParameter(parts(0)) IsNot Nothing Then
                            GetParameter(parts(0)).Value = parts(1)
                        Else
                            If Not Variables.ContainsKey(parts(0)) Then Variables.Add(parts(0), 0)
                            Variables(parts(0)) = _calculator.Evaluate(parts(1))
                        End If
                    End If
                Else
                    _results.Add(lineNumber, String.Empty)
                End If
                lineNumber += 1
            Next

            Return Join(_results.Values.ToArray, Environment.NewLine)
        End Function

        Private Function GetRollFunction() As String
            If RollFunctions.Count = 1 Then Return RollFunctions(0).Value
            Dim rollFunctionParameter As Parameter = GetParameter(ROLL_FUNCTION_PARAMETER)
            If rollFunctionParameter IsNot Nothing Then
                Dim name As Object = rollFunctionParameter.Value
                If name Is Nothing Then
                    Return RollFunctions(0).Value
                Else
                    Return RollFunctions.First(Function(rf As RollFunction) rf.Name = CStr(name)).Value
                End If
            End If
            Return String.Empty
        End Function

        Private Function GetRollResults(ByVal params() As Object) As String
            Dim delimiter As String = " "
            If params.Count > 0 Then delimiter = params(0).ToString
            Return Join(Calculator.LastRollResult.IndividualRolls.ToArray, delimiter)
        End Function

        Private Sub _calculator_EvaluateParameter(name As String, args As NCalc.ParameterArgs) Handles _calculator.EvaluateParameter
            If Variables.ContainsKey(name) Then
                args.Result = Variables(name)
            End If
        End Sub
    End Class
End Namespace

File: Grammars/Dice/DiceRoll.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Xml.Serialization;
using Microsoft.VisualBasic;
using NCalc;

namespace Grammars.Dice
{
    [XmlRoot("DiceRoller", IsNullable = false)]
    public class DiceRoll : BaseGrammar
    {
        private const string ASSIGN_OPERATOR = ":=";
        private const string ROLL_FUNCTION_PARAMETER = "RollFunction";
        private const string COMMENT_CHARACTER = "#";

        private Calculator _calculator;

        private Dictionary<int, string> _results = new Dictionary<int, string>();

        public DiceRoll()
        {
            _calculator = new Calculator(this);
            _calculator.EvaluateParameter += _calculator_EvaluateParameter;
            RollFunctions = new ObservableCollection<RollFunction>();
        }

        [XmlElement("function")]
        public ObservableCollection<RollFunction> RollFunctions { get; set; }

        [XmlElement("supportsMaxLength")]
        public override bool SupportsMaxLength
        {
            get { return false; }
            set { }
        }

        public override string Analyze()
        {
            return string.Empty;
        }

        protected override string GenerateName()
        {
            var lines = new List<string>(Strings.Split(GetRollFunction(), "\n"));
            var lineNumber = 1;

            _results.Clear();

            foreach (var line in lines.Select(l => l.Trim()))
            {
                if (!string.IsNullOrWhiteSpace(line) && !line.StartsWith(COMMENT_CHARACTER))
                {
                    var parts = line.Split(new[] {ASSIGN_OPERATOR}, StringSplitOptions.None);
                    if (parts.Length == 1)
                    {
                        _results.Add(lineNumber, _calculator.Evaluate(line).ToString());
                    }
                    else
                    {
                        if (GetParameter(parts[0]) != null)
                        {
                            GetParameter(parts[0]).Value = parts[1];
                        }
                        else
                        {
                            if (!Variables.ContainsKey(parts[0]))
                            {
                                Variables.Add(parts[0], 0);
                            }
                            Variables[parts[0]] = _calculator.Evaluate(parts[1]);
                        }
                    }
                }
                else
                {
                    _results.Add(lineNumber, string.Empty);
                }
                lineNumber++;
            }

            return string.Join(Environment.NewLine, _results.Values.ToArray());
        }

        private string GetRollFunction()
        {
            if (RollFunctions.Count == 1)
            {
                return RollFunctions[0].Value;
            }
            var rollFunctionParameter = GetParameter(ROLL_FUNCTION_PARAMETER);
            if (rollFunctionParameter != null)
            {
                var name = rollFunctionParameter.Value;
                if (name == null)
                {
                    return RollFunctions[0].Value;
                }
                return RollFunctions.First(rf => rf.Name == (name).ToString()).Value;
            }
            return string.Empty;
        }

        private string GetRollResults(object[] @params)
        {
            var delimiter = " ";
            if (@params.Length > 0)
            {
                delimiter = @params[0].ToString();
            }
            return string.Join(delimiter, Calculator.LastRollResult.IndividualRolls.ToArray());
        }

        private void _calculator_EvaluateParameter(string name, ParameterArgs args)
        {
            if (Variables.ContainsKey(name))
            {
                args.Result = Variables[name];
            }
        }

        public class RollFunction : INotifyPropertyChanged, IDataErrorInfo
        {
            private string _name;
            private string _value;

            [XmlAttribute("name")]
            public string Name
            {
                get { return _name; }
                set
                {
                    if (_name != value)
                    {
                        _name = value;
                        OnPropertyChanged();
                    }
                }
            }

            [XmlText]
            public string Value
            {
                get { return _value; }
                set
                {
                    if (_value != value)
                    {
                        _value = value;
                        OnPropertyChanged();
                    }
                }
            }

            public string Error
            {
                get { return string.Empty; }
            }

            public string this[string columnName]
            {
                get
                {
                    switch (columnName)
                    {
                        case "Name":
                            if (string.IsNullOrWhiteSpace(Name))
                            {
                                return "Name is required.";
                            }
                            break;
                        case "Value":
                            if (string.IsNullOrWhiteSpace(Value))
                            {
                                return "Function is required.";
                            }
                            break;
                    }
                    return string.Empty;
                }
            }

            public event PropertyChangedEventHandler PropertyChanged;

            protected virtual void OnPropertyChanged([CallerMemberName] string propertyName = null)
            {
                if (PropertyChanged != null)
                {
                    PropertyChanged(this, new PropertyChangedEventArgs(propertyName));
                }
            }
        }
    }
}